conn.go

```diff
@@ -218,6 +237,18 @@ func (c *connect) compressBuffer(start int) error {
 }
 
 func (c *connect) sendData(block *proto.Block, name string) error {
+	if c.closed {
+		err := errors.New("attempted sending on closed connection")
+		c.debugf("[send data] err: %v", err)
+		return err
+	}
+
+	if c.buffer == nil {
+		err := errors.New("attempted sending on nil buffer")
+		c.debugf("[send data] err: %v", err)
+		return err
+	}
```

Comment:
yup

Reply:
hmm, this is outdated code, please check the latest changes
https://github.com/ClickHouse/clickhouse-go/pull/1389/files

Reply:
also, safety is not presumed, it's necessary here.
previously, in conn_process.go access to c.close and c.buffer was single-threaded.
now i'm reading data in background goroutine and in foreground thread i'm waiting on select:

``` go
// do reads in background
errCh := make(chan error, 1)
doneCh := make(chan bool, 1)

go func() {
	err := c.processImpl(ctx, on) // this accesses c.reader internally (reads data)
	if err != nil {
		errCh <- err
		return
	}

	doneCh <- true
}()

// select on context or read channel (errors)
select {
case <-ctx.Done():
	c.cancel() // calls c.close(), it accesses c.reader internally (sets c.reader to nil)
	return ctx.Err()

case err := <-errCh:
	return err

case <-doneCh:
	return nil
}
```

1) Closing connections must be done very carefully. Currently **conn.Close()** is broken because multiple calls to **conn.Close()** have issues deadlocking the thread it was called from

2) Also multiple calls to **c.close()** from multiple threads is not thread safe, doing so will result in data races.

So, this needed to be fixed from this:
```
func (c *connect) close() error {
    if c.closed {
        return nil
    }
    c.closed = true
<...>
```

To a thread-safe version of it
```
func (c *connect) close() error {
	c.closeMutex.Lock()
	if c.closed {
		c.closeMutex.Unlock()
		return nil
	}
	c.closed = true
	c.closeMutex.Unlock()
<...>
```

This guarantees that under no circumstances **c.close()** can be called twice no matter what.

3) Besides this we need to guarantee threadsafe access to **c.reader** from **c.processImpl()**, **c.cancel()** and  **c.close()**, hence I've added **c.readerMutex()**

4) One more thing: the order of operations during closing a connection:
Old version:
```
func (c *connect) close() error {
	c.closeMutex.Lock()
	if c.closed {
		c.closeMutex.Unlock()
		return nil
	}
	c.closed = true
	c.closeMutex.Unlock()

	c.buffer = nil

	c.readerMutex.Lock()
	c.reader = nil
	c.readerMutex.Unlock()

	if err := c.conn.Close(); err != nil {
		return err
	}
	
	return nil
}
```

As you can see here, we're changing **c.buffer** and **c.reader** that might be currently reading/sending data, before we close the network connection. Connection must be closed first, in order to stop any i/o operations on **c.buffer** and **c.reader** (and related contexts), then it's safe to deallocale **c.buffer** and **c.reader**.

```
func (c *connect) close() error {
	c.closeMutex.Lock()
	if c.closed {
		c.closeMutex.Unlock()
		return nil
	}
	c.closed = true
	c.closeMutex.Unlock()

	if err := c.conn.Close(); err != nil {
		return err
	}

	c.buffer = nil

	c.readerMutex.Lock()
	c.reader = nil
	c.readerMutex.Unlock()

	
	return nil
}
```